### PR TITLE
move to cfitsio 4

### LINKED ISSUE #334: move to cfitsio 4?
We gotta keep up. I wonder what they changed...

### PR DESCRIPTION
**Pull request by esheldon**
closes #334

### CODE REVIEW

**Comment by esheldon** [comment]
I think the failure is due to these lines being removed from zlib/zconf.h in cfitsio 4

```
84:#  define inflateEnd            z_inflateEnd
417:  #pragma map(inflateEnd,"INEND")
```

**Comment by esheldon** [comment]
actually no, it seems to be because zlib is no longer shipped with cfitsio

**Comment by esheldon** [comment]
This is a significant change to the deps.  Might consider adding the zlib dir from the previous version to this repo.

**Comment by beckermr** [comment]
The external build test is against cfitsio 3. IDK if we went to add 4 or switch?

**Comment by esheldon** [comment]
good catch, will change it

**Comment by esheldon** [comment]
the cfitsio configure says `checking for inflateEnd in -lz... yes`

**Comment by beckermr** [comment]
So I think the issue here is that we need to link in libz directly when building against external builds maybe?

**Comment by beckermr** [comment]
Ugh. Now the bundled build fails. I don't see it building the zlib sources.

File: patches/configure.in.patch [hunk 1]
```diff
@@ -1,31 +1,29 @@
---- cfitsio3470/configure.in	2018-03-01 10:28:51.000000000 -0600
-+++ cfitsio3470/configure.in	2018-12-14 08:39:20.000000000 -0600
-@@ -42,6 +42,13 @@
+--- /home/esheldon/Downloads/cfitsio-4.2.0/configure.in	2022-10-31 14:40:23.000000000 -0400
++++ /home/esheldon/Downloads/cfitsio-4.2.0-mod/configure.in	2023-07-14 11:45:00.797390794 -0400
```
Comment (beckermr) [inline patches/configure.in.patch:2]:
```suggestion
--- cfitsio-4.2.0/configure.in	2022-10-31 14:40:23.000000000 -0400
+++ cfitsio-4.2.0/configure.in	2023-07-14 11:45:00.797390794 -0400
```

File: patches/configure.patch [hunk 2]
```diff
@@ -1,41 +1,38 @@
---- cfitsio3470/configure	2018-03-01 10:28:51.000000000 -0600
-+++ cfitsio3470/configure	2018-12-14 08:39:20.000000000 -0600
-@@ -708,6 +708,7 @@
+--- /home/esheldon/Downloads/cfitsio-4.2.0/configure	2022-10-31 14:40:23.000000000 -0400
++++ /home/esheldon/Downloads/cfitsio-4.2.0-mod/configure	2023-07-14 11:46:53.298055665 -0400
```
Comment (beckermr) [inline patches/configure.patch:2]:
```suggestion
--- cfitsio-4.2.0/configure	2022-10-31 14:40:23.000000000 -0400
+++ cfitsio-4.2.0/configure	2023-07-14 11:46:53.298055665 -0400
```

File: patches/fitscore.c.patch [hunk 3]
```diff
@@ -1,10 +1,11 @@
---- cfitsio3470/fitscore.c	2018-03-01 10:28:51.000000000 -0600
-+++ cfitsio3470/fitscore.c	2018-12-14 08:39:20.000000000 -0600
-@@ -182,6 +182,19 @@
-
+--- /home/esheldon/Downloads/cfitsio-4.2.0/fitscore.c	2022-10-31 14:40:23.000000000 -0400
++++ /home/esheldon/Downloads/cfitsio-4.2.0-mod/fitscore.c	2023-07-14 11:47:54.870417064 -0400
```
Comment (beckermr) [inline patches/fitscore.c.patch:2]:
```suggestion
--- cfitsio-4.2.0/fitscore.c	2022-10-31 14:40:23.000000000 -0400
+++ cfitsio-4.2.0/fitscore.c	2023-07-14 11:47:54.870417064 -0400
```

File: patches/fitsio.h.patch [hunk 4]
```diff
@@ -1,8 +1,8 @@
---- cfitsio3470/fitsio.h	2018-03-01 10:28:51.000000000 -0600
-+++ cfitsio3470/fitsio.h	2018-12-14 08:39:20.000000000 -0600
-@@ -797,6 +797,7 @@
+--- /home/esheldon/Downloads/cfitsio-4.2.0/fitsio.h	2022-10-31 14:40:23.000000000 -0400
++++ /home/esheldon/Downloads/cfitsio-4.2.0-mod/fitsio.h	2023-07-14 11:48:10.102506229 -0400
```
Comment (beckermr) [inline patches/fitsio.h.patch:2]:
```suggestion
--- cfitsio-4.2.0/fitsio.h	2022-10-31 14:40:23.000000000 -0400
+++ cfitsio-4.2.0/fitsio.h	2023-07-14 11:48:10.102506229 -0400
```

File: patches/putcols.c.patch [hunk 5]
```diff
@@ -1,19 +1,20 @@
---- cfitsio3470/putcols.c	2018-03-01 10:28:51.000000000 -0600
-+++ cfitsio3470/putcols.c	2018-12-14 08:39:20.000000000 -0600
-@@ -158,7 +158,16 @@
-
+--- /home/esheldon/Downloads/cfitsio-4.2.0/putcols.c	2022-10-31 14:40:24.000000000 -0400
++++ /home/esheldon/Downloads/cfitsio-4.2.0-mod/putcols.c	2023-07-14 11:49:03.954820768 -0400
```
Comment (beckermr) [inline patches/putcols.c.patch:2]:
```suggestion
--- cfitsio-4.2.0/putcols.c	2022-10-31 14:40:24.000000000 -0400
+++ cfitsio-4.2.0/putcols.c	2023-07-14 11:49:03.954820768 -0400
```

**Review by esheldon — COMMENTED** [review]
this is more clear, but note I don't think patch actually uses those paths

**Comment by esheldon** [comment]
at least the way we are using it where we give the path of the original

**Comment by olebole** [comment]
@esheldon I can't really test that exactly: in Debian, I remove the cfitsio that you provide (to make sure it is not used) and compile/link to the one provided by Debian, by setting the `use_system_fitsio` flag to True. All I can say that it works well with the latest cfitsio version (4.2.0) as a system provided package.

**Comment by beckermr** [comment]
We are asking if the build flag you use (`--use-system-fitsio`) still works on this branch. We had to change how it is implemented due to backend changes in pip and setuptools. Can you test a debian rpm build with the code on this branch?

**Comment by beckermr** [comment]
lol jinx

**Comment by olebole** [comment]
Ah, OK. I didn't check this yet. May take a few days until I find time; is this OK?

**Comment by esheldon** [comment]
few days is fine, thank you.

**Comment by olebole** [comment]
I checked this, and can confirm that it works for me by adding the `--use-system-fitsio` option to `setup.py`.

**Comment by esheldon** [comment]
Thanks @olebole

**Pull request merged**